selfdrive/test/process_replay/regen.py

```diff
@@ -117,48 +117,100 @@ def replay_service(s, msgs):
       rk.keep_time()
 
 
-def replay_cameras(lr, frs, disable_tqdm=False):
+def replay_single_camera(s, stream, dt, vipc_server, frames, size):
+  services = [(s, stream)]
+  pm = messaging.PubMaster([s for s, _ in services])
+  rk = Ratekeeper(1 / dt, print_delay_threshold=None)
+
+  img = b"\x00" * int(size[0] * size[1] * 3 / 2)
+  while True:
+    if frames is not None:
+      img = frames[rk.frame % len(frames)]
+
+    rk.keep_time()
+
+    m = messaging.new_message(s)
+    msg = getattr(m, s)
+    msg.frameId = rk.frame
+    msg.timestampSof = m.logMonoTime
+    msg.timestampEof = m.logMonoTime
+    pm.send(s, m)
+
+    vipc_server.send(stream, img, msg.frameId, msg.timestampSof, msg.timestampEof)
+
+  
+def replay_dual_camera(keys, streams, frames, sizes, dt, vipc_server):
+  if len(keys) != len(streams) != len(frames) != len(sizes) != 2:
+    raise ValueError("keys, streams, frames, and sizes must all be length 2")
```

Comment:
why is this a different function than `replay_single_camera`?

Reply:
To synchronize sending of fcam and ecam messages. I figured that if they share the same timestamps and are sent sequentially in the same process, the modeld receives them in order and is actually using matching frames for eval. Otherwise you get a lot of messages about frames being out of sync.